File: Part 2 - Regression/Polynomial Regression/polynomial_regression.R
# Bluffing Detector
# Polynomial Linear Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')

dataset = dataset[2:3]

# Fitting Linear Regression Model to the dataset
lin_reg = lm(formula = Salary ~ ., data = dataset)

# Fitting the Polynomial Regression Model to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ ., data = dataset)


library(ggplot2)

# Visualizing the Linear Regression Results
ggplot() +
	geom_point(aes(x = dataset$Level , y = dataset$Salary),
			   colour = 'red') +
	geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
			  colour = 'blue') +
	ggtitle("Truth or Bluff? (Linear Regression)") +
	xlab("Level") +
	ylab("Salary")

# Visualizing the Polynominal Linear Regression Results
ggplot() +
	geom_point(aes(x = dataset$Level , y = dataset$Salary),
			   colour = 'red') +
	geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
			  colour = 'blue') +
	ggtitle("Truth or Bluff? (Polynomial Linear Regression)") +
	xlab("Level") +
	ylab("Salary")

# Visualizing the Polynominal Linear Regression Results (High Resolution)
X_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
dataset_high_resol = data.frame(Level = X_grid,
								Level2 = X_grid^2,
								Level3 = X_grid^3,
								Level4 = X_grid^4)
ggplot() +
	geom_point(aes(x = dataset$Level , y = dataset$Salary),
			   colour = 'red') +
	geom_line(aes(x = X_grid, y = predict(poly_reg, newdata = dataset_high_resol)),
			  colour = 'blue') +
	ggtitle("Truth or Bluff? (Polynomial Linear Regression - High Resol)") +
	xlab("Level") +
	ylab("Salary")

# Predicting the new result with Linear Regression Model
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

# Predicting the new result with Linear Regression Model
poly_new_data = data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4)
y_pred = predict(poly_reg, newdata = poly_new_data)

File: Part 2 - Regression/Polynomial Regression/polynomial_regression.py
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 13 11:24:05 2019

@author: Niharika_Cnr
"""

# Bluffing Detector using Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X, y)

#Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures

poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X)

poly_lin_reg = LinearRegression()
poly_lin_reg.fit(X_poly, y)

#---------------------------------------------------------------------------------------#
## Visualising Linear Regression Results
#plt.scatter(X, y, color = 'r')
#plt.plot(X, lin_reg.predict(X), color = 'b')
#plt.title("Truth or Bluff? (Linear Regression)")
#plt.xlabel("Position Level")
#plt.ylabel("Salary")
#plt.show()
#
## Visualizing Polynomial Regression Results
#X_grid = np.arange(min(X), max(X), 0.1)
#X_grid = X_grid.reshape(len(X_grid), 1)
#plt.scatter(X, y, color = 'r')
#plt.plot(X_grid, poly_lin_reg.predict(poly_reg.fit_transform(X_grid)), color = 'b')
#plt.title("Truth or Bluff? (Polynomial Regression)")
#plt.xlabel("Position Level")
#plt.ylabel("Salary")
#plt.show()
#---------------------------------------------------------------------------------------#

# Predicting a new result with Linear Regression Model
lin_reg.predict([[6.5]])

# Prediciting new result with Polynominal Regression Model
poly_lin_reg.predict(poly_reg.fit_transform([[6.5]]))
